### ISSUE TITLE
provide map/filter-type operations for string-dicts and other collections

### ISSUE THREAD

**Issue by shriram** [brownplt/pyret-lang#935]
That's all. Feature request.

**Comment by jswrenn** [comment]
There are a few different signatures that `map` could have on a string-dict:

1. `(A → B), StringDict<A> → StringDict<B>`
2. `({String;A} → B), StringDict<A> → StringDict<B>`
3. `({String;A} → {String;B}), StringDict<A> → StringDict<B>`

I've included that last one for completeness, but I think it's needlessly complex.

Second: should `map` on a `MutableStringDict` perform differently than `map` on a `StringDict`?

**Comment by ds26gte** [comment]
The closest prior art I can find is Common Lisp's `maphash` which has the signature `{String, A -> B}, StringDict<A>} -> void`. I.e., the function is only called for side-effect. Since we probably would prefer something functional, especially for immutable string dicts, your #2 seems the best bet.

For mutables, it seems attractive to have the argument string-dict (`MutStringDict<A>`) be modified, but that only makes sense if the incoming function is of type `A -> A`, which may not be reasonable. We could just keep the resulting string-dict be also mutable.

**Comment by ds26gte** [comment]
Implemented in ds26gte/pyret-lang@e5aceaa.

To zoom into the change, click the "3 changed files" link: `string-dict.js` contains the implementation, `test-string-dict.arr` contains the tests.

Summary. Given an immutable-string-dict `isd` and a mutable-string-dict `msd`, the following hold:

* `isd.map(f]` produces an immutable-string-dict that has the same keys as `isd`, but each key now points to `f(a)` where it originally pointed to `a`
* `isd.filter(p)` produces an immutable-string-dict that is a subdict of `isd`, retaining only those keys for whose values `p` is true
* `msd.map(f)` is like the immutable-string-dict version, except its result is mutable
* `msd.filter(p)` side-effects `msd` itself, retaining only the pertinent keys. It returns `nothing`.

It would seem that `msd.map(f)` should also side-effect `msd`, but that would be type-correct only if `f`'s argument and result type are identical, which would be too restrictive.

If a different semantics would be desirable, let me know.

**Comment by shriram** [comment]
Thanks for the suggestions, Jack, and quick turn-around, Dorai.

I'm not a fan of these alternate definitions for the name `map`. I think it
would be terribly confusing for `map` to be different on different
datatypes.

This will also preclude us from eventually getting to some kind of
"implements mappable" interface, which we may well do (and which we
informally have in the form of methods). If we proceed with that anyway,
it'll break a bunch of code, and that code will not be easy to find and fix
because we don't have static types. And then we'd still be stuck with what
to call this other function.

Also, if we're talking about conventions from other languages, I'm not sure
CL is the right guide. Modern collection-rich languages seem to all have
the "standard" type for `map`. It may nevertheless be useful to see what
Scala and F# do.

=====

I think there's a basic problem in talking about "mapping" over an SD.
Mapping conventionally makes sense when you have a linear structure with a
single value in it. But we have an (artificially-ordered) structure with
two distinguished values. The RHS (mappee) is not always more important
than the LHS (mapper: the string key).

I expect that the _conventional_ interpretation of `map` is still to
rewrite the RHS values, just as one expects `map` on lists to NOT rewrite
the list indices (i.e., to re-order the list). For that reason, and for the
reasons above, the function we should associate with `map` must have the
type #1.

But the other two are not only valid, they're even useful. I think this
utility is _much greater_ when we talk about `filter`, because it is not at
all clear that filtering must always proceed on the RHS than on the LHS. I
can see situations in which I'd want to filter by the keys, not even
looking at their associated value.

Nevertheless, by the same reasoning above, I believe `filter` should
process single elements, which should be the RHS. But we need a `filter`
that operates on the LHS (key) as well, as well as one that takes both.

=====

Based on this, and for completeness, I suggest the following names and
types:

```
  map :: (A -> B), SD<A> -> SD<B>
  map-v = map
  map-kv :: (String, A -> B), SD<A> -> SD<B>  # Jack, not sure why the
tuple, when arity 2 works
  map-kvk :: (String, A -> {String; B}), SD<A> -> SD<B> # with keys updated
```

I've left out `map-k :: (String -> String), SD<A> -> SD<A>` and maybe
others since they can be built using `map-kvk`. There's an argument for
leaving `map-kv` in place since it is a pleasant interface, likely to be
useful, general, while avoiding having to return a tuple value. Likewise:

```
  filter :: (A -> Boolean), SD<A> -> SD<A>
  filter-v = filter
  filter-k :: (String -> Boolean), SD<A> -> SD<A>
  filter-kv :: (String, A -> Boolean), SD<A> -> SD<A>
```

I think this covers just about all the cases?

=====

All `map` and `filter` operations should produce new collections, not
perform in-place mutation. This should be so _even_ if the initial
collection is mutable. It is reasonable to have mutable parallels of each
of the above operations, but by default in Pyret we always produce new
values.

The mutability of the new collection should mirror that of its input.

=====

@ds26gte, are these being implemented atop the new immutable string
library? I presume so?

**Comment by jpolitz** [comment]
Implementation comment – see the use of `runtime.eachLoop` in https://github.com/brownplt/pyret-lang/blob/master/src/js/trove/str-dict.js#L772.

This is (a) stack-safe, and (b) faster than doing recursive `safeCall`.

**Comment by blerner** [comment]
`map-kvk` is underdefined: what should it do when the mapping function causes a collision among keys?  Since iteration order is unspecified, this function would also have to be unspecified...  Also, its signature is asymmetric: it takes two arguments but returns a tuple.  A cleaner signature would be Jack's signature 3 above, which maps entries to entries.

I don't particularly like this function, and don't think it's common enough to warrant building in.

Also, for completeness, we'll need `each-k`, `each-kv`, and `each-v`...

**Comment by jpolitz** [comment]
Two other helpers to consider that have been _extremely_ useful on lists, especially for avoiding multiple traversals:

```
partition<K, V> :: (V -> Bool) -> { SD<K, V>; SD<K, V> }
```

The first element of the tuple is all the "true" responses, second is all the "false" responses.




```
filter-map<K, A> :: (A -> Option<B>) -> SD<K, B>
```

(All the `none` return values are removed, all the `some` values are kept)

**Comment by shriram** [comment]
Oh yeah, for sure, totally agree with Joe's suggestions. One question: why
not make `partition` return an object with fields named `true` and `false`?
The main reason I see is that we then lose destructuring assignment (in the
current setting, where we can't do it on fields), though being able to
`.true` and `.false` means there's not so much of a win to destructuring
assignment in the first place. [Maybe better names than those of built-in
constants: `matched` and `not-matched`, or something like that.]

Basically, why not make the return value self-documenting instead of having
to rely on positional memory?

**Comment by blerner** [comment]
True and false are keywords...

**Comment by jpolitz** [comment]
@shriram compare your opinion here to https://github.com/brownplt/pyret-lang/issues/797#issuecomment-236449849

What changed?

**Comment by shriram** [comment]
Joe, good point. I think Ben's response has good counter-arguments.

**Comment by ds26gte** [comment]
Second attempt with @shriram's suggested quintet (`map map-kv map-kvk filter filter-kv`) at ds26gte/pyret-lang@be169aa.

Incorporated @jpolitz's suggestions about ensuring stack safety. Also this commit does not re-bootstrap. Only two files modified: `string-dict.js` (and it is based on the new immutable library) and the test file `test-string-dict.arr`.

**Comment by jpolitz** [comment]
![screen shot 2016-11-09 at 6 27 46 pm](https://cloud.githubusercontent.com/assets/119891/20158730/418c4856-a6aa-11e6-96a1-1f59b284cc24.png)

I'm seeing three changed files, including phase0/pyret.jarr

**Comment by ds26gte** [comment]
That's still the old commit. Browser cache? The new commit id should be `be169aa` (ds26gte/pyret-lang@be169aa). I'd created the branch `local-master` anew (to avoid having to keep around the unnecessary `pyret.jarr` changes) and re-pushed it.

![screenshot from 2016-11-09 22-59-49](https://cloud.githubusercontent.com/assets/3708729/20164153/5f95917e-a6d0-11e6-8f59-15afba96ed43.png)

**Comment by jpolitz** [comment]
Got it, I thought you were referring to the first one.

**Comment by ds26gte** [comment]
@blerner: To confirm, your suggested `each...` are like the `map...` but are used purely for side-effect? (I.e., they return `nothing`?)

**Comment by blerner** [comment]
Yes

**Comment by ds26gte** [comment]
ds26gte/pyret-lang@f281fd7  now includes `filter-k` (missed it in the previous round) plus @blerner's suggested `each`, `each-k`, and `each-kv`.
